workers/src/verifier.rs

```diff
@@ -0,0 +1,152 @@
+use alloy::providers::{Provider, ProviderBuilder, ReqwestProvider};
+use reqwest::{Client, ClientBuilder, Url};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, std::fmt::Debug)]
+pub(crate) struct ResponseWrapper {
+    pub message: Message,
+}
+
+#[derive(Serialize, Deserialize, std::fmt::Debug)]
+pub(crate) struct Message {
+    #[serde(rename = "RollupId")]
+    rollup_id: u32,
+    #[serde(rename = "BlockHeight")]
+    block_height: u64,
+    #[serde(rename = "Timestamp")]
+    timestamp: u64,
+    #[serde(rename = "NearDaTransactionId")]
+    near_da_transaction_id: Vec<u8>,
+    #[serde(rename = "NearDaCommitment")]
+    near_da_commitment: Vec<u8>,
+    #[serde(rename = "StateRoot")]
+    pub(crate) state_root: Vec<u8>,
+}
+
+// TODO: Generify in a future for other networks, like Solana.
+/// Verifies state roots for the Decentralized Verification Network (DVN).
+///
+/// This verifier implements the V1 verification algorithm which:
+/// 1. Fetches the state root from the aggregator
+/// 2. Retrieves the block state root using the chain RPC API
+/// 3. Compares the two state roots to determine message validity
+pub struct NFFLVerifier {
+    http_client: Client,
+    eth_l2_provider: ReqwestProvider<Ethereum>,
+    aggregator_http_address: String,
+    network_id: String,
+}
+
+impl NFFLVerifier {
+    /// For test purposes only
+    pub(crate) async fn new(agg_url: &str, eth_l2_url: &str, network_id: u64) -> eyre::Result<NFFLVerifier> {
+        let client = ClientBuilder::new().build()?;
+        let url: Url = eth_l2_url.parse()?;
+        let provider = ProviderBuilder::new().on_http(url);
+        let agg_http_addr = format!("{}/state-root", agg_url);
+
+        Ok(NFFLVerifier {
+            eth_l2_provider: provider,
+            http_client: client,
+            aggregator_http_address: agg_http_addr,
+            network_id: network_id.to_string(),
+        })
+    }
+
+    pub async fn new_from_config(cfg: &DVNConfig) -> eyre::Result<NFFLVerifier> {
+        Self::new(&cfg.aggregator_url, &cfg.http_rpc_url, cfg.network_eid).await
+    }
+
+    /// Verifies the state root of a block. In case if any request future
+    /// is interrupted, or finishes unsuccessfully, returns Ok(false).
+    pub async fn verify(&self, block_height: u64) -> eyre::Result<bool> {
+        match tokio::try_join!(
+            self.get_aggregator_root_state(block_height),
+            self.get_block_state_root(block_height),
+        ) {
+            Ok((agg_response, block_state_root)) => {
+                let state_root_slice: &[u8] = agg_response.message.state_root.as_slice();
+                let aggregator_state_root: B256 = B256::from_slice(state_root_slice);
+                if agg_response.message.block_height != block_height {
+                    return Ok(false);
+                }
+                Ok(block_state_root.eq(&aggregator_state_root))
+            }
+            Err(_) => Ok(false),
```

Comment:
Hm. I did it intentionally because on my understanding, if you can't fetch data from at least one source -> you are not able to validate it -> it is enough to return false. But yeah, I'll add log message.